### PR TITLE
SecureRandom enhancements

### PR DESCRIPTION
**Pull request by ysbaddaden**
Implements SecureRandom.uuid generator and the convenient SecureRandom.base64 and SecureRandom.urlsafe_base64 for quick and secure token generations.

Examples:

```crystal
SecureRandom.uuid           # "36b98a7e-888a-445d-88b0-47a390b52f33"
SecureRandom.base64         # "YtG7cYH191c4177HJIsxEQ=="
SecureRandom.urlsafe_base64 # "gMNU_WXUG2LmNMb8_1QpXQ"
```

### CODE REVIEW

**Comment by bararchy** [comment]
@ysbaddaden Great job ! we needed that at Amethyst too :) :+1:

File: src/secure_random.cr [hunk 1]
```diff
@@ -19,4 +29,16 @@ module SecureRandom
...
+  def self.uuid
+    bytes = random_bytes(16)
+    {
+      bytes[0, 4].hexstring,
+      bytes[4, 2].hexstring,
+      bytes[6, 2].hexstring,
+      bytes[8, 2].hexstring,
+      bytes[10, 2].hexstring,
+      bytes[12, 4].hexstring,
+    }.join("-")
```
Comment (ysbaddaden) [inline src/secure_random.cr]: This is weird but Tuple#join(String) was faster than Tuple#join(Char).

**Comment by will** [comment]
Version 4 UUIDs have the form xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx where x is any hexadecimal digit and y is one of 8, 9, A, or B (e.g., f47ac10b-58cc-4372-a567-0e02b2c3d479).

File: src/secure_random.cr (hunk 1 above)
Comment (asterite) [inline src/secure_random.cr]: I tried this:

```crystal
require "benchmark"

TUP = {"a", "b", "c", "d", "e", "f", "g"}

Benchmark.ips do |x|
  x.report("string") do
    TUP.join "-"
  end
  x.report("char") do
    TUP.join '-'
  end
end
```

and get:

```
string 3476001.50 (± 23731.54)  1.17× slower
  char 4076697.63 (± 27724.80)       fastest
```

In any case you are generating many intermediate strings there (the `hexstring` calls). The best thing would be to have a separate overload for `hexstring` accepting an IO, then using Enumerable#join with IO and a block.

Comment (ysbaddaden) [inline src/secure_random.cr]: I ran this benchmark on linux64 and got opposite results:
```
string 1267725.75 (± 108942.84)       fastest
  char 1222427.06 (± 84191.20)  1.04× slower
```

Comment (jhass) [inline src/secure_random.cr]:
```
string 3402285.49 (± 484877.03)  1.19× slower
  char 4052882.12 (± 673219.84)       fastest

Linux jhass-laptop 4.0.5-1-ARCH #1 SMP PREEMPT Sat Jun 6 18:37:49 CEST 2015 x86_64 GNU/Linux
```
for me.

**Comment by ysbaddaden** [comment]
Thanks @will! I totally overlooked that!

**Comment by ysbaddaden** [comment]
I now deal with a single buffer in a dumb, but efficient and clean way. This should be ~1.4× faster.

I'll squash commits as necessary when we're good :)

File: src/secure_random.cr [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 module SecureRandom
+  def self.base64(n = 16)
+    Base64.encode64(random_bytes(n)).delete('\n')
+  end
+
+  def self.urlsafe_base64(n = 16, padding = false)
+    str = base64(n).tr("+/", "_")
+    padding ? str : str.delete('=')
+  end
+
```
Comment (kostya) [inline src/secure_random.cr]: isn't both of this methods implemented in base64 module?

Comment (ysbaddaden) [inline src/secure_random.cr]: Intents are differents: Base64 encodes data when SecureRandom returns a token of a given size. I agree those 2 methods are merely sugar for `encode64(random_bytes(n))`.

But you're right: there is a urlsafe_encode64 that I should be using!

Comment (kostya) [inline src/secure_random.cr]: scrict_encode64 also

File: src/int.cr [hunk 3]
```diff
@@ -254,6 +254,10 @@ struct Int
     end
   end
 
+  def to_hex
```
Comment (jhass) [inline src/int.cr]: I'm not sure I fully follow what this does, but I don't think to_hex is a good name for a public API for it. Hexadecimal is about representation, but this doesn't return a representation, it returns a new value in the current representation (the `Int` subclass, which is also a representation of a number). From the name I would expect it to return the `Char`s `'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'` (or their lowercase counterparts) and raise for values > 15.

Comment (ysbaddaden) [inline src/int.cr]: I moved it here to reduce some duplication, but this isn't needed anymore. I'm not sure there is any need for this method in the public API.

**Comment by ysbaddaden** [comment]
Based on everybody's comments here comes an overall rework/squash:

- leverages existing base64 strict and urlsafe methods
- generates correct UUIDv4
- faster UUID implementation using a single String buffer
- doesn't introduce Int#to_hex

File: src/secure_random.cr (hunk 1 above)
Comment (will) [inline src/secure_random.cr]: So on all of these, the std dev is _huge_ (±673219.84, ±108942.84, etc) which makes either one coming ahead in the margin of error.

**Comment by asterite** [comment]
@ysbaddaden Thanks! Looks really good.

One question: about `def hexstring(buffer, offset = 0)`, what do you think if we remove the offset? Because on the caller side you can do `buffer + offset` (so in a way that `offset` argument is redundant).

**Comment by ysbaddaden** [comment]
Yes, if you can avoid the offset param, that would be great. I wasn't sure how to.

**Comment by asterite** [comment]
Ouch, totally forgot to merge this before 0.7.5. Sorry!

**Pull request merged**